internal/wal_verify_handler.go

```diff
@@ -0,0 +1,219 @@
+	}
+}
+
+// QueryCurrentWalSegment() gets start WAL segment from Postgres cluster
+func QueryCurrentWalSegment() WalSegmentDescription {
+	conn, err := Connect()
+	tracelog.ErrorLogger.FatalfOnError("Failed to establish a connection to Postgres cluster %v", err)
+
+	queryRunner, err := newPgQueryRunner(conn)
+	tracelog.ErrorLogger.FatalfOnError("Failed to initialize PgQueryRunner %v", err)
+
+	currentSegmentNo, err := getCurrentWalSegmentNo(queryRunner)
+	tracelog.ErrorLogger.FatalfOnError("Failed to get current WAL segment number %v", err)
+
+	currentTimeline, err := getCurrentTimeline(conn)
+	tracelog.ErrorLogger.FatalfOnError("Failed to get current timeline %v", err)
+
+	err = conn.Close()
+	tracelog.WarningLogger.PrintOnError(err)
+
+	// currentSegment is the current WAL segment of the cluster
+	return WalSegmentDescription{Timeline: currentTimeline, Number: currentSegmentNo}
+}
+
+// HandleWalVerify queries the current cluster WAL segment and timeline
+// and travels through WAL segments in storage in reversed chronological order (starting from that segment)
+// to find any missing WAL segments that could potentially fail the PITR procedure
+func HandleWalVerify(rootFolder storage.Folder, startWalSegment WalSegmentDescription, outputWriter WalVerifyOutputWriter) {
+	walFolder := rootFolder.GetSubFolder(utility.WalPath)
+	storageFileNames, err := getFolderFilenames(walFolder)
+	tracelog.ErrorLogger.FatalfOnError("Failed to get WAL folder filenames %v", err)
+
+	storageSegments := getSegmentsFromFiles(storageFileNames)
+	timelineSwitchMap, err := createTimelineSwitchMap(startWalSegment.Timeline, walFolder)
+	tracelog.ErrorLogger.FatalfOnError("Failed to initialize timeline history map %v", err)
+	walSegmentRunner := NewWalSegmentRunner(startWalSegment, storageSegments, 1, timelineSwitchMap)
+
+	// maxConcurrency is needed to determine max amount of missing WAL segments
+	// after the last found WAL segment which can be skipped ("uploading" segment sequence size)
+	maxConcurrency, err := getMaxUploadConcurrency()
+	tracelog.ErrorLogger.FatalOnError(err)
+
+	segmentScanner := NewWalSegmentScanner(walSegmentRunner)
+	err = runWalIntegrityScan(segmentScanner, maxConcurrency)
+	tracelog.ErrorLogger.FatalfOnError("Failed to perform WAL segments scan %v", err)
+
+	integrityScanSegmentSequences, err := collapseSegmentsByStatusAndTimeline(segmentScanner.ScannedSegments)
+	tracelog.ErrorLogger.FatalOnError(err)
+
+	err = outputWriter.Write(newWalVerifyResult(integrityScanSegmentSequences))
+	tracelog.ErrorLogger.FatalOnError(err)
+}
+
+// runWalIntegrityScan invokes the following storage scan series:
+// 1. At first, it runs scan until it finds some segment in WAL storage
+// and marks all encountered missing segments as "missing, probably delayed"
+// 2. Then it scans exactly uploadingSegmentRangeSize count of segments,
+// if found any missing segments it marks them as "missing, probably still uploading"
```

Comment:
Here we use maxUploadCOncurrency as uploadingSegmentRangeSize, for example, it will be 2.

So we will mark 7 and 8 as Uploading, and 5,6 will have the status delayed. But 5,6 appears earlier and probable these segments should have Uploading status instead of 7,8.

Reply:
In this case will mark all of them as delayed I guess. Uploading segment range size may contain or may not contain some segments, it is started exactly after some segment has been found on storage
I draw some illustration, forgot to pin to the PR
![image](https://user-images.githubusercontent.com/47750602/91407837-4629c800-e85c-11ea-9e53-7bc73d7d8ad0.png)

Reply:
So red (missing) segments in the first group on the picture are MISSING_DELAYED, missing in the second group are  MISSING_UPLOADING, missing in third are MISSING_LOST

Reply:
I've added more docs in https://github.com/wal-g/wal-g/pull/746/commits/d4de3fe15bce6ac9281221cee14facc172c85a20